Add adapters - Base design + 'structures' (+ 'references'... sort of)

CasperWA opened the pull request:
The goal of the adapters is to contain a conversion or adaptation ability in a single Python class.

The class will be a proxy for `optimade.models.EntryResource`, where the specific entries will take the place, e.g., `StructureResource`.

In this way, one can instatiate the adapter class with a JSON resource object matching it, and get all the properties as attributes. E.g., one could get the lattice vectors of a structure then by calling
`structures_adapter_class_object.attributes.lattice_vectors`. The reason to not have the adapter class be a direct link or subclass of the model class, is to not inherit all that comes with it, i.e., do not inherit `pyndantic`'s `BaseModel`, etc. While it can trip most IDEs, it is a more elegant solution that opens the doors to other possibilities, in my opinion.

Besides being a handy wrapper for an entity resource object for a client, it also serves as a converter/adapter to other known and widely used libraries for the specific entity type. For a structures resource object, this may be libraries such as pymatgen, ASE, AiiDA, etc...
It may also be nice to be able to convert the structures resource object to a Crystallographic Information File (CIF) file or other. These adapter/conversion functions may be added in a separate file and then referenced in the adapter class.

Thus, the adapter class stays small at first glance, but spreads its tentacles wide throughout both the current package, but also a multitude of other libraries, packages, and data formats.

In general, whenever the adapter class tries to get an attribute, it goes through the following steps:

1. If the attribute starts with `get_`, try to look up if this is a known conversion, then perform the conversion and return it as a string or specific package/library Python class.
2. Try to fetch the OPTIMADE model property.
3. Raise AttributeError if the attribute is still unknown, i.e., the previous steps failed and didn't raise already.

---

Missing parts of this PR:

- Add tests:
  - [X] General tests for the adapter class.
  - Tests for each converter function:
    - [x] AiiDA
    - [x] ASE
    - [x] CIF
    - [x] PDB
    - [x] PDBx/mmCIF
    - [x] pymatgen
- [x] Generalize the adapter main class and have `Structure` inherit from it.
- [x] Add `Reference`.
  (Currently added as an empty inheritence, meaning there are currently no conversions available for `Reference`).

Helpful additions/corrections to this PR:

- Optimize the various conversion functions (especially `get_cif` and `get_pbdx_mmcif`).
- Add more well-known or much used data format converters and packages/libraries.
  (Keeping in mind that some of these libraries have a large quantity of converters already, e.g., ASE already converts to PDB and CIF formats, and so therefore these should only be included if going through, e.g., ASE, may end in a loss of data - I may not have been entirely strict enough with this myself already 😅).

ltalirz commented on optimade/adapters/structures/pymatgen.py:
```diff
@@ -0,0 +1,89 @@
+__all__ = ("get_pymatgen_structure",)
+
+
+def get_pymatgen_structure(
```
I guess the consistent name here would be `get_pymatgen ?

ltalirz replied:
I see you have a different convention for `ase` (`get_ase_atoms`)  vs pdb (`get_pdb`)

CasperWA replied:
Agreed with `get_pymatgen`. Maybe even if we should follow the convetion for the others (i.e., AiiDA and ASE) it should be `get_pymatgen_structure_molecule`. I like the shorter, more concise version, better though.

Yup. I think I'll change `get_ase_atoms` to `get_ase`.

For AiIDA, I think it's still fine to keep the extra explicit mentioning of the Node in question, since it can be expanded to include `get_aiida_cifnode` or something as well...?

CasperWA replied:
> I see you have a different convention for `ase` (`get_ase_atoms`) vs pdb (`get_pdb`)

For all of these, the function name that converts may differ from the actual format to be specified in the `get_` attribute. So one must indeed specify `get_pymatgen` to get a pymatgen Structure or Molecule, as well as one must specify `get_ase` to get an ASE Atoms instance.
But you want the conversion function to reflect the attribute name?

ltalirz commented on optimade/adapters/structures/__init__.py:
```diff
@@ -0,0 +1,89 @@
+        """Get converted structure or attribute from OPTIMADE structure
+        Order:
+        - Try to return converted structure if using `get_<valid_structures value>`.
+        - Try to return OPTIMADE StructureResource attribute.
```
You are mixing namespaces here... is this really necessary?
What about defining a `def convert(self, format)` method that does the job?

CasperWA replied:
Nice idea with `convert()`.
Concerning the latter part - I'm not sure exactly what you mean? Would you have me sub-class `optimade.models.StructureResource`? I wanted explicitly _not_ to do this, in order to avoid getting the `pydantic` things and validators that followed along with it. But maybe it's fine, I'm not sure?

ltalirz replied:
Sorry for not being clearer. I simply meant to keep the second part of this function, which returns attributes of `self.structure`.

ltalirz commented on optimade/adapters/structures/__init__.py:
```diff
@@ -0,0 +1,89 @@
+        "pdb": get_pdb,
+        "pdbx_mmcif": get_pdbx_mmcif,
+        "pymatgen": get_pymatgen_structure,
+    }
```
Eventually, this might become a `ConverterCollection` class that could include checks on the interface of the functions it contains etc.
For the moment, this is fine

CasperWA replied:
Great idea. And definitely one for the future, i.e., not something I will explore at this time.

shyamd commented:
I'm not fully seeing what this does. How does this plug into the server for instance?

CasperWA commented:
> I'm not fully seeing what this does. How does this plug into the server for instance?

It's not for server use, more for client use. The main goal is to interpret OPTIMADE structures and retrieve their information/convert them to client-usable data formats and constructs.

CasperWA commented:
> thanks @CasperWA , I think having these converters in the optimade python tools is very useful.

Thanks :) Me too, hence the "fold back" from the client I'm working on.

> May I suggest to add both individual extras for the corresponding dependencies + a `converters` extra that contains them all?

I thought about this, got tired at the idea, and set it aside for a time when I have not more exciting ideas for how the adapter should work. I.e., I will definitely add this to the `setup.py` 👍

ltalirz commented on optimade/adapters/structures/__init__.py:
```diff
@@ -0,0 +1,89 @@
+            return self._converted_structures[structure_type]
+
+        # Try returning StructureResource attribute
+        try:
```
I think what you want here is simply
```
return self.structure.__getattr__(name)
```

CasperWA replied:
Yes and no.
Yes, because it is indeed what I want, and am getting the same result with `getattr()`.
No, because I also want to display my specialized `AttributeError` message :) In the end, I don't see the difference between `getattr()` and `__getattr__()` in this case - although the latter may be prettier and slightly clearer? Or is there a true benefit to one over the other do you think?

CasperWA replied:
Hmm. Thinking more about this, I hesitate. Mainly because we're using internal Python class methods here, and I'd rather go the "proper" way so any `pydantic` logic isn't corrupted or something other - especially thinking of the difference between `__getattribute__()` and `__getattr__()`, where `pydantic` may have implemented something clever for these methods. I could of course just go the Python way and call `__getattribute__()`, but why not just use the more common `getattr()` function?

ltalirz replied:
[Here](https://stackoverflow.com/a/1944708/1069467) is a good explanation of the difference.

Now that you're moving out the conversion to `convert`, I'm not sure you still need a custom error message.
The default one should already be quite adequate.

CasperWA replied:
> [Here](https://stackoverflow.com/a/1944708/1069467) is a good explanation of the difference.

I looked into the difference as well - which is why I changed my mind, since `pydantic` is surely playing around with these methods. And to be on the safe side, I'd rather go through the proper "Python API" ... :)

> After looking at this, I'm no longer sure what I suggested is indeed the correct solution.
> Your solution with `getattr` might be better.
> 
> However, now that you're moving out the conversion to `convert`, I'm not sure you still need a custom error message.
> The default one should already be quite adequate.

You sure? Having an extra helpful message concerning the available formats and how to get them is not useful?

ml-evs commented:
Not going to review this until everyone else has finished hacking at it (ping me @CasperWA), but think this looks very useful and is definitely in the spirit of "optimade-python-tools". 

Once we get this in, I think we should focus for a while on writing docs that showcase how this repo can be used by implementers of servers and clients. Would we even want to provide a simple Python/CLI client implementation in this repo, given that we have most of the tools already? (I guess this would depend on how much this treads on the toes of your fancier client @CasperWA)

CasperWA commented:
@ml-evs @ltalirz I consider this ready for review now.
I have added a lot of tests, and tried to come up with edge-cases, and what to do for various optional properties. This has hopefully resulted in more robust conversion functions.

Some of the conversion function tests are the same from test file to test file. This may be made more elegant, but in order to properly separate out the test functions if one should not have the specific dependency installed, this is the design I chose.

Note also, the kind of ugly addition to our CI workflows, due to AiiDA.
In order to create a `StructureData` AiiDA Node, one _must_ have a functioning AiiDA profile loaded, hence all the extra stuff.
I have opened an issue aiidateam/aiida-core#3898 to try and make this more gentle on the lines of code in our workflows. Since this design idea from the side of AiiDA is so deeply integrated into its code base, it is difficult to see how we can get to a state, where it may not be needed to set up an AiiDA profile, just for testing this.

Finally, if you have any optimizations for the various converters, now is the time! :) If you want to add a bunch of new ones (especially to the `references` adapter), perhaps consider either doing it in another PR or _quickly_ add it to this PR :)
I know the PDBx/mmCIF is a bit of a mess, but I don't have the time now to research a whole new and to me unknown format to get it properly up to speed - sorry. I feel this is better than nothing for now?

CasperWA commented:
> Not going to review this until everyone else has finished hacking at it (ping me @CasperWA), but think this looks very useful and is definitely in the spirit of "optimade-python-tools".

Consider yourself "pinged" @ml-evs :) (But not until Tuesday the 14th and onwards).

> Once we get this in, I think we should focus for a while on writing docs that showcase how this repo can be used by implementers of servers and clients. (...)

I think that would definitely be the next big "project" for the repo. Also, we should make sure to update to v1.0.0 ASAP (both the package version, but I was thinking mainly of the OPTIMADE API version). Maybe v1 of the package should indeed wait until we have docs in place.

> (...) Would we even want to provide a simple Python/CLI client implementation in this repo, given that we have most of the tools already? (I guess this would depend on how much this treads on the toes of your fancier client @CasperWA)

So as soon as this PR goes through, we could (in theory) adopt my client. That way I can update the current version with _this_ adapter. And I don't know how fancy it is... But you can now [check it out on binder](https://mybinder.org/v2/gh/aiidalab/aiidalab-optimade/pure_jupyter_version?urlpath=%2Fapps%2FOPTIMADE_general.ipynb) to judge for yourself 😅 
I could even move the repository [aiidalab/aiidalab-optimade](https://github.com/aiidalab/aiidalab-optimade) over to Materials-Consortia? I have discussed this loosely with @giovannipizzi already.
I would still want to create an AiiDA lab-specific edition of the client, but the most recent version, which is about to be merged into the `develop` branch, is completely AiiDA- and AiiDA lab-agnostic - only depending on the Jupyter environment, really.

But I don't know if we want an OPTIMADE client in Jupyter under the Materials-Consortia umbrella?

In any case! We should probably create an issue for this discussion, or rather pick it up at a consortium meeting.

shyamd commented:
I would say we want to make a stark distinction between a client and a jupyter-widget. The client should be agnostic of use, so generally just a python class that can be instantiated with an OPTIMADE endpoint URL that allows traversal and retrieval. I would argue this `adapters` PR is really a part of the client and not core. We might think about breaking this into 3 namespace-packages. One for core models and definitions. One for the server. One for the client.

ml-evs commented:
> Consider yourself "pinged" @ml-evs :) (But not until Tuesday the 14th and onwards).

:+1:

> I think that would definitely be the next big "project" for the repo. Also, we should make sure to update to v1.0.0 ASAP (both the package version, but I was thinking mainly of the OPTIMADE API version). Maybe v1 of the package should indeed wait until we have docs in place.

:+1:

> So as soon as this PR goes through, we could (in theory) adopt my client. That way I can update the current version with _this_ adapter. And I don't know how fancy it is... But you can now [check it out on binder](https://mybinder.org/v2/gh/aiidalab/aiidalab-optimade/pure_jupyter_version?urlpath=%2Fapps%2FOPTIMADE_general.ipynb) to judge for yourself sweat_smile

This is very impressive, I love it! Definitely going to have a look through the code, I didn't realise ipywidgets could be so powerful. 

This goes far above and beyond what I had in mind for this repository, but perhaps there is potential for some overlap? I was imagining some simple client classes that allow you to get the pydantic models from a server in pure Python (trivial given that we do this already) and perhaps a small CLI on top of that. This could obviously be achieved with `wget`... perhaps we could think of a simple OPTIMADE use case that could be achieved more easily with this approach, e.g. use the CLI to find all structures of a given formula across all of the providers and stick them in a json file, with an easy way to read it back in with the pydantic models.

> In any case! We should probably create an issue for this discussion, or rather pick it up at a consortium meeting.

:+1:

ml-evs commented:
> I would say we want to make a stark distinction between a client and a jupyter-widget. The client should be agnostic of use, so generally just a python class that can be instantiated with an OPTIMADE endpoint URL that allows traversal and retrieval. I would argue this `adapters` PR is really a part of the client and not core. We might think about breaking this into 3 namespace-packages. One for core models and definitions. One for the server. One for the client.

I think we're in agreement, though I'm not against bundling the models/definitions with our example server and a lightweight example client in one package (for now, at least) but we should discuss this at the next meeting.

We currently have some entangled layers to unpick, like the server being required to generate an openapi spec (not sure we can get around this), and I don't think our grammar/filtering is mature enough yet to be separated from the example server either, at least not before OPTIMADE 1.0.

shyamd commented:
Actually that's exactly what namespace packages are for. We can still develop these packages under the same repo. As a result, build-time cross dependencies are not an issue. Since all the packages will share one version, we can keep them all consistent as well.

https://packaging.python.org/guides/packaging-namespace-packages/

For an example on how this works, you might look at our MPContribs repo:
https://github.com/materialsproject/MPContribs

ml-evs commented:
> Actually that's exactly what namespace packages are for. We can still develop these packages under the same repo. As a result, build-time cross dependencies are not an issue. Since all the packages will share one version, we can keep them all consistent as well.

Ah, gotcha, this looks like a sensible way to go then.

ml-evs commented on optimade/adapters/base.py:
```diff
@@ -0,0 +1,75 @@
+
+        # Try returning ENTRY_RESOURCE attribute
+        try:
+            res = getattr(self.entry, name)
```
I see in the tests that this is intended, so no bother.

CasperWA replied:
Yeah, sorry, the `attribute` in the comment line may be confusing here. The intention was just return every property at the toplevel of `EntityResource`, i.e., `id`, `type`, `attributes`, etc.
So if one actually _wants_ a value under `attributes` it would, e.g., go like `entity.attributes.nsites`.

CasperWA replied:
Later on we could add more fanciful detection, to first check at the top-level, and later under `attributes` and maybe even under `relationships` or something? So that `entity.nsites` works alongside `entity.id`.

ml-evs replied:
> Later on we could add more fanciful detection, to first check at the top-level, and later under `attributes` and maybe even under `relationships` or something? So that `entity.nsites` works alongside `entity.id`.

That's what I thought you were aiming for from the comment (maybe just tweak the wording?), could be useful in the future, I agree.

CasperWA replied:
Maybe I could just add it now? Although, I would like to add some proper testing for this as well.. hmm... I'll see what happens.

ml-evs commented on tests/adapters/structures/test_ase.py:
```diff
@@ -0,0 +1,72 @@
+        Periodicity.PERIODIC,
+        Periodicity.APERIODIC,
+    )
+    structure.attributes.lattice_vectors = (
```
Any way we could generalise these tests a bit?

e.g. some wrapper test that could be called as  `_wrap_test_null_lattice_vectors(get_ase_atoms, Atoms)` which always acts on the modified data defined in `structure`?

CasperWA replied:
Okay, so I've restructured the testing quite a bit now.
As it stands, it may even be able to simplify it even more.
It would be good to have "automagic" testing of any new conversion function added to the repository, alongside with specific manually written tests. I may do that in the next iteration.

CasperWA commented on .github/workflows/deps_eager.yml:
```diff
@@ -58,7 +84,7 @@ jobs:
 
   #   steps:
-  #   - uses: actions/checkout@master
+  #   - uses: actions/checkout@v2
```
I guess we could

CasperWA replied:
I have created issue #247 to track this.

CasperWA commented on .github/workflows/requirements.txt:
```diff
@@ -1,5 +1,5 @@
-lark-parser==0.8.1
-fastapi==0.52.0
+lark-parser==0.8.5
+fastapi==0.53.1
```
I don't think Dependabot sees these files, only the `setup.py` file.
At least I didn't add these files, since I wasn't sure if that was wise. I can try though?

ml-evs replied:
It looks like it updated these files in #219, hence my confusion. I think we should be updating these versions to track the highest supported version in `setup.py`.

CasperWA replied:
I see your confusion. I believe those edits were mine, on top of the dependabot PR.

CasperWA replied:
Created issue #249 to track this.
I will set this up _after_ this PR.

CasperWA commented on optimade/adapters/base.py:
```diff
@@ -0,0 +1,75 @@
+
+class EntryAdapter:
+    """Base class for lazy resource entry adapters
+    :param entry: JSON OPTIMADE single resource entry.
```
Yeah, this is not easy as long as we don't have docs in place yet for this particular repository. As soon as that's in place, we should unify the pydoc style.

CasperWA commented on optimade/adapters/base.py:
```diff
@@ -0,0 +1,75 @@
+            "If you want to get a converted entry use `get_<entry_type>`, "
+            f"where `<entry_type>` is one of {tuple(self._type_converters.keys())}\n"
+            "Otherwise, you can try to retrieve an OPTIMADE "
+            f"{self.ENTRY_RESOURCE.__class__.__name__} attribute."
```
It indeed should. Don't know why it didn't. Did you instantiate `adapter = EntryAdapter()`? Because this shouldn't be done. One should _always_ go through the subclassed ones. I.e., the `EntryAdapter` class shouldn't be part of the public API.

ml-evs replied:
```python
import requests
from optimade.adapters import Structure
doc = requests.get('https://odbx.science/optimade/v0/structures/odbx/1').json()["data"]
adapter = Structure(doc)
print(adapter.xxx)
```

yields

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mevans/src/optimade/optimade-python-tools/optimade/adapters/base.py", line 71, in __getattr__
    f"Unknown attribute: {name}.\n"
AttributeError: Unknown attribute: xxx.
If you want to get a converted entry use `get_<entry_type>`, where `<entry_type>` is one of ('aiida_structuredata', 'ase', 'cif', 'pdb', 'pdbx_mmcif', 'pymatgen')
Otherwise, you can try to retrieve an OPTIMADE ModelMetaclass attribute.
```

CasperWA replied:
Ah. Huh. I'll have to try and see what I can do about that.

CasperWA commented on optimade/adapters/base.py:
```diff
@@ -0,0 +1,75 @@
+    """
+
+    ENTRY_RESOURCE: EntryResource = EntryResource
+    _type_converters: Dict[str, Callable] = {}
```
I think this may very easily be done by patching through to the underlying `pydantic` model? But I don't know if that's desireable as such, since we don't have direct control over what is and isn't included that way.

CasperWA replied:
In the end, this would simply return the original input, wouldn't it?

ml-evs replied:
It would, and I think that would be useful as the input is easily lost otherwise (unless you go into `adpater._entry...` etc. The adapter is also implicitly validating the JSON you pass it as an OPTIMADE resource so there is some useful work being done.

CasperWA replied:
True. Okay - I'll add it in as a converter. Perhaps it should be a valid converter for all adapters? So it may be implemented in the parent.

CasperWA replied:
I've now added common converters in the base adapter for `json` and `dict`, i.e., one can call `entry.get_json` or `entry.get_dict` and it should work.
For now there is no logic in place to check that `json` and `dict` are not added to `_type_converters`.

CasperWA commented on optimade/adapters/references/__init__.py:
```diff
@@ -0,0 +1,17 @@
+    """
+
+    ENTRY_RESOURCE = ReferenceResource
+    _type_converters = {}
```
I was thinking of this, but didn't want to throw myself into researching bibjson.
Instead, I was hoping someone with the knowledge of it would add a converter, either to this PR or afterwards - and to any other type of desired conversion format.
However, I can give a short look into it now, and if it takes too long, I'll make an issue for it once this is merged or ping you if you want to give it a go before.

ml-evs replied:
I think this ties in with the comment above about a standard `json` "converter"

CasperWA commented on optimade/adapters/structures/ase.py:
```diff
@@ -0,0 +1,53 @@
+    atoms = []
+    for site_number in range(attributes.nsites):
+        species_name = attributes.species_at_sites[site_number]
+        site = attributes.cartesian_site_positions[site_number]
```
It seemed to me to be so. I've also added some tests for these cases, which you should check out. I believe some of them include tests of converting structures with `null` coordinates?

CasperWA commented on optimade/adapters/structures/pymatgen.py:
```diff
@@ -0,0 +1,105 @@
+        warn(PYMATGEN_NOT_FOUND)
+        return None
+
+    # pymatgen cannot handle unknown positions
+    for site in optimade_structure.attributes.cartesian_site_positions:
+        if None in site:
+            raise ConversionError(
+                "pymatgen cannot be used to convert structures with unknown positions."
+            )
```
Sounds like a valid design choice.
I mostly added these after the fact, i.e., after having written the tests and realizing they failed for `null` coordinates. So it's indeed just a form of hacks in each function. Something that should be made better in hindsight.

CasperWA commented on optimade/server/config.py:
```diff
@@ -15,6 +16,7 @@
 
 
 DEFAULT_CONFIG_FILE_PATH = str(Path.home().joinpath(".optimade.json"))
+logger = logging.getLogger("optimade")
```
Agreed.

CasperWA replied:
I've created issue #248 to keep track of this.

CasperWA commented on setup.py:
```diff
@@ -4,19 +4,30 @@
 # Dependencies
-mongo_deps = ["pymongo~=3.10", "mongomock~=3.19"]
-server_deps = ["uvicorn", "Jinja2~=2.11"] + mongo_deps
+# Server minded
```
Yeah. It would be great to see an outline of this.

CasperWA commented on tests/adapters/structures/test_aiida.py:
```diff
@@ -0,0 +1,89 @@
+    )
+
+
+def test_null_lattice_vectors():
```
Although, how would you ever receive such a `lattice_vector` value? That's not OPTIMADE compliant as far as I know? It should always be a 3 long list of 3 long lists, i.e., `((None, None, None), (1, 1, 1), (1, 1, 1))`? Do you think we should try and treat cases that the OPTIMADE spec says are not allowed as well? Just curious, as I'm having a bit of trouble deciding how far these adapters should go in error handling.

ml-evs replied:
Ah I must have mixed up the lattice vector discussion in my head, for some reason I thought we were allowing `None` instead of `(None, None, None)`. Feel free to resolve any other comments on this...

CasperWA commented on optimade/adapters/structures/aiida.py:
```diff
@@ -0,0 +1,85 @@
+    attributes = optimade_structure.attributes
+
+    # Handle any None values in lattice_vectors (turn Null into 1.0)
+    lattice_vectors = str(attributes.lattice_vectors)
+    adjust_cell = False
+    if "None" in lattice_vectors:
+        adjust_cell = True
+        lattice_vectors = lattice_vectors.replace("None", "1.0")
+    lattice_vectors = ast.literal_eval(lattice_vectors)
```
Nodding to another comment, I don't think we need to take care of the issue where an inner vector is simply a single `None`, however, I like this approach better than my "convert to `str`, handle it there, convert back", which seemed to me to be a bit fragile from the get to.

CasperWA commented on optimade/adapters/structures/pymatgen.py:
```diff
@@ -0,0 +1,105 @@
+    attributes = optimade_structure.attributes
+
+    return Structure(
+        lattice=attributes.lattice_vectors,
```
Pymatgen doesn't return a `pymatgen.Structure`, but rather a `pymatgen.Molecule` for OPTIMADE structures with unknown lattice vectors (assuming the `dimension_types` are set accordingly). `pymatgen.Molecule` doesn't use `lattice_vectors` for whatever reason, I guess since it doesn't actually define a structure.
If you know more about pymatgen, you're welcome to correct it, so that we don't unneccesarily lose information during the conversion that we could've kept.

CasperWA commented:
@ml-evs I haven't addressed all your comments yet, and I don't know how much more I will be able to do today.
However, it is in a much better state now, I think. The tests have been modularized and a clear path to doing some centralized testing of all conversion functions has opened up, but not yet followed.
Even more, is that it seems most of the modules _can_ indeed handle unknown values, as long as it's explicitly converted to `float("nan")`.

CasperWA commented:
All right @ml-evs ! I have finished addressing your review comments, it should be ready for re-review.

ml-evs commented on optimade/adapters/base.py:
```diff
@@ -0,0 +1,115 @@
+        - Try to return OPTIMADE ENTRY_RESOURCE.attributes attribute.
+        - Raise AttributeError
+        """
+        # get_<entry_type>
```
Minor quibble with this `adapter.get_format` nomenclature is that I would expect e.g. `get_json` to be a callable, do you think something like `adapter.as_json` is a better name?

CasperWA replied:
Yeah, that might be a good idea. I feel the same way as you about `get`, so `as` seems the better choice.

ml-evs commented on optimade/adapters/base.py:
```diff
@@ -0,0 +1,115 @@
+            return res
+
+        try:
+            res = getattr(self.entry.attributes, name)
```
It's just occurred to me that this ends up polluting the adapter namespace with all the methods of a pydantic model (rather than just its data). We could fix this by using something like:

```suggestion
        if name in self.entry.attributes.__field_set__:
            return getattr(self.entry.attributes, name)
```

which only looks at fields set on initializ=sation of the pydantic object (could also use just `__fields__` but that seems to include some pydantic stuff like `immutable_id`).

CasperWA replied:
`immutable_id` _is_ an OPTIMADE property.

CasperWA replied:
I think `__fields_set__` returns a set of the fields that were actually also set. So a double "set" of sorts ...
While `__fields__` returns the complete model, i.e., all available properties/attributes, not excluding even the ones that weren't set when instantiating the given model with some data set.

ml-evs replied:
oops, hopefully it was just me misreading the `__fields_set__` then, it must contain it

ml-evs commented on tests/adapters/references/test_references.py:
```diff
@@ -0,0 +1,86 @@
+            new_Reference = Reference(reference)
+            assert isinstance(new_Reference.entry, ReferenceResource)
+
+    def test_setting_entry(self, capfd, RAW_REFERENCES):
```
TIL some new pytest magic :)

CasperWA replied:
[Fixtures](https://docs.pytest.org/en/latest/fixture.html) for the win :)

ml-evs commented on optimade/adapters/structures/proteindatabank.py:
```diff
@@ -0,0 +1,241 @@
+            f"CRYST1{cellpar[0]:9.3f}{cellpar[1]:9.3f}{cellpar[2]:9.3f}"
+            f"{cellpar[3]:7.2f}{cellpar[4]:7.2f}{cellpar[5]:7.2f} P 1\n"
+        )
+
```
I've had some issues (again with VESTA) getting the positions read in the correct format. I think if you supply a `CRYST` block, some readers/viewers assume the positions are fractional. We can fix this by adding [`SCALEn` blocks](https://www.wwpdb.org/documentation/file-format-content/format33/sect8.html#SCALEn) that contain the transformation back to fractionals. If we're happy to use numpy here, this would be (seems to ned the extra 0):

```suggestion

         vol = np.dot(cellpar[0], np.cross(cellpar[1], cellpar[2]))
         for i, vec in enumerate(cellpar):
                scale = np.cross(cellpar[i + 1 % 3], cellpar[i + 2 % 3]) 
                pdb += (
                    f"SCALE{i} {scale[0]:9.3f}{scale[1]:9.3f}{scale[2]:9.3f} 0.00000"
                )   
```

ml-evs commented on optimade/adapters/structures/cif.py:
```diff
@@ -0,0 +1,129 @@
+            "  'x, y, z'\n\n"
+        )
+
+    # Is it a periodic system?
+    # NOTE: This is a bit ahead of its time, since this OPTIMADE property is part of an open PR.
+    # See https://github.com/Materials-Consortia/OPTIMADE/pull/206
+    coord_type = (
+        "fract" if hasattr(attributes, "fractional_site_positions") else "Cartn"
+    )
```
To summarise our slack discussion, I think the best thing to do is convert to fractionals here if a cell is provided, and if not just output the cartesian positions as you would for a molecule

ml-evs replied:
Actually not even sure we can write a cif file without a cell... maybe we just raise an error?

CasperWA replied:
> Actually not even sure we can write a cif file without a cell... maybe we just raise an error?

I can't see why we shouldn't be able to do this. It will instead be a problem for the viewer in question whether they can handle cif files without cell information or not.

However, for the other issue, I agree (as well from the slack discussion) that we should convert to fractional coordinates if we have sufficient information to do so and otherwise convert to a "molecule-style" output.

CasperWA replied:
Actually, I am already _not_ including cell information in the cif file if we are not dealing with a 3D structure.

ml-evs commented on optimade/adapters/structures/utils.py:
```diff
@@ -15,6 +15,32 @@
+
+    fractional = np.linalg.solve(cell.T, cartesian_positions.T).T
+
+    # Expecting a bulk 3D structure here, note, this may change in the future.
+    # See `ase.atoms:Atoms.get_scaled_positions()` for ideas on how to handle lower dimensional structures.
+    # Furthermore, according to ASE we need to modulo 1.0 twice.
```
Huh, turns out `-1e-35 % 1.0 == 1.0`.

CasperWA replied:
Right, that's why it should be twice! I'll write it in here.
